src/algebra/field.lean

```diff
@@ -5,46 +5,95 @@ Authors: Robert Lewis, Leonardo de Moura, Johannes Hölzl, Mario Carneiro
+
+* `division_semiring` is a nontrivial `semiring` with multiplicative inverses for nonzero elements;
+* `division_ring` is a nontrivial `ring` with multiplicative inverses for nonzero elements,
+  e.g., quaternions;
+* `semifield` is a commutative `division_semiring`, e.g., non-negative real numbers `ℝ≥0`;
+* `field` is a commutative `division_ring`, e.g., `ℚ`, `ℝ`, `ℂ`, `ℤ_[p]` etc.
+
+## Tags
+
+division ring, field
+-/
 open set
 
 set_option old_structure_cmd true
 
 universe u
 variables {K : Type u}
 
-/-- A `division_ring` is a `ring` with multiplicative inverses for nonzero elements -/
-@[protect_proj, ancestor ring has_inv]
-class division_ring (K : Type u) extends ring K, div_inv_monoid K, nontrivial K :=
+/-- A `division_semiring` is a `semiring` with multiplicative inverses for nonzero elements. -/
+@[protect_proj, ancestor semiring div_iv_monoid nontrivial]
+class division_semiring (K : Type u) extends semiring K, div_inv_monoid K, nontrivial K :=
 (mul_inv_cancel : ∀ {a : K}, a ≠ 0 → a * a⁻¹ = 1)
 (inv_zero : (0 : K)⁻¹ = 0)
 
-section division_ring
-variables [division_ring K] {a b : K}
+section division_semiring
 
-/-- Every division ring is a `group_with_zero`. -/
+variables [division_semiring K] {a b : K}
+
+/-- Every division semiring is a `group_with_zero`. -/
 @[priority 100] -- see Note [lower instance priority]
-instance division_ring.to_group_with_zero :
+instance division_semiring.to_group_with_zero :
   group_with_zero K :=
-{ .. ‹division_ring K›,
-  .. (infer_instance : semiring K) }
+{ .. ‹division_semiring K› }
 
 lemma inverse_eq_has_inv : (ring.inverse : K → K) = has_inv.inv :=
 begin
   ext x,
   by_cases hx : x = 0,
   { simp [hx] },
-  { exact ring.inverse_unit (units.mk0 x hx) }
+  { exact (units.mk0 x hx).inverse_eq }
 end
 
+@[field_simps] lemma div_add_div_same (a b c : K) : a / c + b / c = (a + b) / c :=
+by simpa only [div_eq_mul_inv] using (add_mul a b (c⁻¹)).symm
+
+lemma add_div (a b c : K) : (a + b) / c = a / c + b / c :=
+(div_add_div_same _ _ _).symm
+
+lemma one_div_mul_add_mul_one_div_eq_one_div_add_one_div (ha : a ≠ 0) (hb : b ≠ 0) :
+  (1 / a) * (a + b) * (1 / b) = 1 / a + 1 / b :=
+by rw [mul_add, one_div_mul_cancel ha, add_mul, one_mul, mul_assoc, mul_one_div_cancel hb, mul_one,
+  add_comm]
+
+lemma add_div_eq_mul_add_div (a b : K) {c : K} (hc : c ≠ 0) : a + b / c = (a * c + b) / c :=
+(eq_div_iff_mul_eq hc).2 $ by rw [right_distrib, (div_mul_cancel _ hc)]
+
+@[field_simps] lemma add_div' (a b c : K) (hc : c ≠ 0) : b + a / c = (b * c + a) / c :=
+by rw [add_div, mul_div_cancel _ hc]
+
+@[field_simps] lemma div_add' (a b c : K) (hc : c ≠ 0) : a / c + b = (a + b * c) / c :=
+by rwa [add_comm, add_div', add_comm]
+
+end division_semiring
+
+/-- A `division_ring` is a `ring` with multiplicative inverses for nonzero elements -/
+@[protect_proj, ancestor ring div_iv_monoid nontrivial]
+class division_ring (K : Type u) extends ring K, div_inv_monoid K, nontrivial K :=
```

Comment:
If you make this
```suggestion
class division_ring (K : Type u) extends ring K, division_semiring K :=
```
then you don't need the instance below

Reply:
This is bad because it introduces the redundant fields `zero_mul` and `mul_zero` (which in a ring you can prove by distributivity of addition).